Record Teachers' Session Count and IP History

Reimirno opened the pull request.

Reimirno commented:
No client-facing (or admin facing) changes so no cucumber tests added. Should we add rspec test like now, or wait until we modify the teacher table admin view to show the session_count field then add cucumber tests instead?

Reimirno commented:
And, I chose to use `inet` array instead of `cidr` array because it seems to store more info with the same data length. And honestly I don't think there is a big difference between the two. Also inet seems more permissive.

From official doc:
[The essential difference between inet and cidr data types is that inet accepts values with nonzero bits to the right of the netmask, whereas cidr does not. For example, 192.168. 0.1/24 is valid for inet but not for cidr.... Note that if you want to accept only networks, you should use the cidr type rather than inet.](https://www.postgresql.org/docs/current/datatype-net-types.html#:~:text=Note%20that%20if%20you%20want,of%20bits%20in%20the%20netmask.)

Also see this https://stackoverflow.com/questions/73824616/what-is-the-ideal-datatype-to-be-used-to-store-ip-address-in-rails-postgresql-ap

Reimirno reviewed (commented):
migration result of another branch (for the page model) somehow appeared here lol (but only the model and schema files are changed, the migration file per se are not here.
Not sure whether this is by design or did I do something super wrong. But I can see if this is by design it can avoid conflict when merging the branches together. In the past I had serious problem with merging two branches that both touched migration stuff...

cycomachead commented:
> No client-facing (or admin facing) changes so no cucumber tests added. Should we add rspec test like now, or wait until we modify the teacher table admin view to show the session_count field then add cucumber tests instead?

I think the rspec tests are still good. 
In this case we have `controller` tests which can assert things about session count. 
I am not sure what the IP will look like in test mode, but it might be good to have a test for that.

Additionally, it could be useful to have as test where as an admin, when I update another teacher's record, the admin's IP is NOT logged.

cycomachead reviewed (changes requested):
Nice, this looks great! 

Could we add a couple of controller specs. 
Also  for the ip logging let's add a method to the teacher _model_, `append_ip_address`, and this method can prevent appending duplicates. This could also be the place were we remove items in the future if we want to. then we can write `@teacher.append_ip_address(request.remote_ip)`

Thank you!

cycomachead commented:
@Reimirno Also one thing I noticed about this PR:
You have the migrations for this PR AND the schema updates for other PR combined into one. So, there's an unexpected dependency here. 

If you wanted to, you could make 1 new PR which extracts just the data migrations and applies the schema updates. Then we can merge that in really easily. :) I'd also be happy to merge that into the golden repo.

Reimirno commented on app/controllers/teachers_controller.rb:
```diff
@@ -66,6 +66,7 @@ def create
     end
 
     @teacher = Teacher.new(teacher_params)
+    @teacher.session_count += 1
```
Increment session count on sign up

Reimirno commented:
Removed all irrelevant schema footprints. Schema changes related to this ticket still retained (tell me if that's not what you want).

Added Rspec tests. I half-googled and half-pattern-matched and half-github-copiloted but eventually I pulled it off... and hopefully they are correct and idiomatic.

Streamline the logic a bit (removed duplicates as you suggested, added in the missing logic - I forgot to handle the sign up case earlier)
Now:
- on sign up: session+1, ip recorded
- on log in: session +1, ip recorded
- on update (if not admin): ip recorded
